### PR TITLE
*: enable postgres stash

### PR DESCRIPTION
**Pull request by mjibson**
Enable a postgres stash in coord, and plumb it through from the binary and into testdrive for catalog verification.

### Motivation

  * This PR adds a known-desirable feature. #12207

### Testing

- [x] This PR has adequate test coverage / QA involvement has been duly considered.

### Release notes

This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):

  - n/a

### CODE REVIEW

**Comment by mjibson** [comment]
TLS postgres is untested and probably doesn't work. How do the testdrive changes look here?

File: src/materialized/src/bin/materialized/main.rs [hunk 1]
```diff
@@ -359,6 +359,9 @@ pub struct Args {
     /// directory.
     #[clap(long, env = "MZ_PERSIST_CONSENSUS_URL")]
     persist_consensus_url: Option<Url>,
+    /// Postgres catalog stash connection string.
+    #[clap(long, env = "MZ_CATALOG_POSTGRES_STASH", value_name = "POSTGRES_URL")]
```
Comment (philip-stoev) [inline src/materialized/src/bin/materialized/main.rs:363]: How is this postgres URL going to look like in production? Are we going to be using any randomly-generated passwords or just TLS? I think for security reasons, the password should not be part of the URL and should not be part of the command line either. It should be read from a kubernetes secret if running in a kubernetes environment.

**Comment by philip-stoev** [comment]
@mjibson I would like to spend some time on this, being such a fundamental change.

File: src/materialized/src/bin/materialized/main.rs (hunk 1 above)
Comment (mjibson) [reply src/materialized/src/bin/materialized/main.rs:363]: I'm happy to support whatever the cloud team wants to do here, but don't have many opinions myself. Thinking through the security of the postgres url, connection, and how that gets plumbed down can happen from discussion with them.

**Comment by philip-stoev** [comment]
Item No 1 In `services.py` we need to pin postgres to the right default version to be used , otherwise it will pick "latest" which changes over time.

File: src/materialized/src/bin/materialized/main.rs (hunk 1 above)
Comment (benesch) [reply src/materialized/src/bin/materialized/main.rs:363]: The cloud team can plumb the URL securely via the `MZ_CATALOG_POSTGRES_STASH` environment variable!

File: src/testdrive/src/action/sql.rs [hunk 2]
```diff
@@ -183,19 +184,62 @@ impl Action for SqlAction {
...
+                        .batch_execute(
+                            "
+                                CREATE TABLE fence AS SELECT * FROM public.fence;
+                                CREATE TABLE collections AS SELECT * FROM public.collections;
+                                CREATE TABLE data AS SELECT * FROM public.data;
+                                CREATE TABLE sinces AS SELECT * FROM public.sinces;
+                                CREATE TABLE uppers AS SELECT * FROM public.uppers;
+                            ",
+                        )
+                        .await?;
```
Comment (benesch) [inline src/testdrive/src/action/sql.rs:222]: Haha, nice. I might drop a comment here that says "TODO: consider adding a read-only API to stash directly" but these seems like exactly the right solution for right now.

Comment (benesch) [inline src/testdrive/src/action/sql.rs:222]: Haha, nice. I might drop a comment here that says "TODO: consider adding a read-only API to stash directly" but these seems like exactly the right solution for right now.

**Comment by philip-stoev** [comment]
Item No 2. The stuff in the stash is not garbage collected on `DROP`, right? So it remains there forever. I think this is a problem, especially in the context of temporary tables, which still cause rows to be written to the `data` table. If your workload uses a lot of temporary tables this can grow unbounded.

**Comment by philip-stoev** [comment]
Item No 4. The `time` column in the `data` table does not look right:

```
postgres=# select time, count(*) from data group by time limit 10;
         time         | count 
----------------------+-------
 -9223372036854775318 |     3
 -9223372036854774576 |     3
 -9223372036854775430 |     3
 -9223372036854774418 |     3
 -9223372036854775221 |     3
 -9223372036854775356 |     3
 -9223372036854774326 |     3
 -9223372036854775754 |     3
 -9223372036854775438 |     3
 -9223372036854775466 |     3
```

So time seems to be increasing but yet it is a very low negative value ?

**Comment by mjibson** [comment]
> Item No 1 In services.py we need to pin postgres to the right default version to be used , otherwise it will pick "latest" which changes over time.

What's the correct way to do this? There doesn't seem to be well established precedent.

mysql specifies a specific version: https://github.com/MaterializeInc/materialize/blob/6b4e249abe2e6928a00b4c8467cc98318f2feb4c/misc/python/materialize/mzcompose/services.py#L316

postgres comes with an extra mzbuild thing that does something:
https://github.com/MaterializeInc/materialize/blob/6b4e249abe2e6928a00b4c8467cc98318f2feb4c/misc/python/materialize/mzcompose/services.py#L343

sqlserver doesn't specify a version:
https://github.com/MaterializeInc/materialize/blob/6b4e249abe2e6928a00b4c8467cc98318f2feb4c/misc/python/materialize/mzcompose/services.py#L364

debezium allows configuring via some variable:
https://github.com/MaterializeInc/materialize/blob/6b4e249abe2e6928a00b4c8467cc98318f2feb4c/misc/python/materialize/mzcompose/services.py#L387

I'm not sure exactly what I should do here. Seems like I can't just change `image` in `class Postgres` to be something because of the mzbuild stuff that does something. Should I specify this outside of the class __init__?

**Comment by philip-stoev** [comment]
@mjibson I think we need a `DEFAULT_POSTGRES_VERSION` constant in services.py and then follow the template used for the Confluent services.

**Comment by mjibson** [comment]
> Item No 4. The time column in the data table does not look right:

Those timestamps look right to me. They do indeed start at i64::MIN and go up from there. The timestamps there are not regular mz timestamps, but stash timestamps.

**Comment by philip-stoev** [comment]
Item No 5. There seems to be some quadratic or cubic algorithm kicking in because after I created a couple of thousand tables with a script like this:

```
cols = ", ".join(f"f{i} INTEGER" for i in range (100))
for i in range(10000):
    print(f"CREATE TABLE t{i} ({cols});");
```

materialized hangs on restart and is never able to get to the state where it accepts connection. There is one thread that is spinning like this:

```
Thread 1 (Thread 0x7fd84f71a840 (LWP 2106034)):
#0  0x0000560df954be52 in alloc::collections::btree::node::Handle<alloc::collections::btree::node::NodeRef<alloc::collections::btree::node::marker::Mut,K,V,alloc::collections::btree::node::marker::Leaf>,alloc::collections::btree::node::marker::Edge>::insert_recursing ()
#1  0x0000560df9ed6bd3 in alloc::collections::btree::map::BTreeMap<K,V>::insert ()
#2  0x0000560df961c643 in mz_stash::TableTransaction<K,V,I>::verify ()
#3  0x0000560df96220b7 in mz_stash::TableTransaction<K,V,I>::update ()
#4  0x0000560df9b84fa9 in mz_coord::catalog::storage::Transaction<S>::update_item ()
#5  0x0000560df95a0f26 in mz_coord::catalog::migrate::rewrite_items ()
#6  0x0000560df97c529d in mz_coord::catalog::Catalog<S>::open::{{closure}} ()
#7  0x0000560df96c9e29 in materialized::serve_stash::{{closure}} ()
#8  0x0000560df9a79c41 in <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll ()
#9  0x0000560df9d055fd in std::thread::local::LocalKey<T>::with ()
#10 0x0000560df9e6d432 in tokio::park::thread::CachedParkThread::block_on ()
#11 0x0000560df9df30d8 in tokio::runtime::thread_pool::ThreadPool::block_on ()
#12 0x0000560df9a18d91 in tokio::runtime::Runtime::block_on ()
#13 0x0000560df9c1ea6a in materialized::run ()
#14 0x0000560df9c1adee in materialized::main ()
#15 0x0000560df9df1eb3 in std::sys_common::backtrace::__rust_begin_short_backtrace ()
#16 0x0000560df9fafe89 in std::rt::lang_start::{{closure}} ()
#17 0x0000560dfbcfc1d1 in core::ops::function::impls::<impl core::ops::function::FnOnce<A> for &F>::call_once () at /rustc/7737e0b5c4103216d6fd8cf941b7ab9bdbaace7c/library/core/src/ops/function.rs:259
#18 std::panicking::try::do_call () at library/std/src/panicking.rs:492
#19 std::panicking::try () at library/std/src/panicking.rs:456
#20 std::panic::catch_unwind () at library/std/src/panic.rs:137
#21 std::rt::lang_start_internal::{{closure}} () at library/std/src/rt.rs:128
#22 std::panicking::try::do_call () at library/std/src/panicking.rs:492
#23 std::panicking::try () at library/std/src/panicking.rs:456
#24 std::panic::catch_unwind () at library/std/src/panic.rs:137
#25 std::rt::lang_start_internal () at library/std/src/rt.rs:128
#26 0x0000560df9c76b12 in main ()
```

**Comment by philip-stoev** [comment]
Item No 6. In a multi-homed scenario, are we going to use separate databases or separate schemas for the separate Mz instances? I am asking because you are not allowed to specify a schema in a tokio_postgres connection URL, so it just defaults to `public', which is assumed to exist
```
pstoev@Ubuntu-2004-focal-64-minimal:~/mjibson-coord-stash2$ target/release/materialized --catalog-postgres-stash postgres://postgres:postgres@localhost
materialized: stash error: db error: ERROR: no schema has been selected to create in
```

One can specify `?dbname=xxxx` in the URL though.

**Comment by philip-stoev** [comment]
Item No 7. This statement:

```
materialize=> ALTER INDEX i1 SET (logical_compaction_window = '50000ms');
ALTER INDEX
```
does not produce any rows in the data table, contrary to e.g. `ALTER RENAME`, which produces a retraction + insertion, as expected.

**Comment by mjibson** [comment]
> Item No 6. In a multi-homed scenario, are we going to use separate databases or separate schemas for the separate Mz instances

My understanding (also a cloud decision) is separate databases, so that we can have separate users for each database to prevent possible cross tenant data access.

**Comment by philip-stoev** [comment]
Item No 7. I may be raising universal stash issues rather than postgres-specific ones, but I did not test the sqlite stash when it landed ...

So in the data table, the time appears to increment individually per collection id. Is that expected? Even if a DDL operation that will create a record with `collection_id = 5` was the last that was ever run, there are rows in the table with other `collection_id`s whose `time` is higher.

**Comment by philip-stoev** [comment]
Item No 9. We do not seem to be to tolerate any postgres downtime. Mz will panic immediately instead of retrying:

```
2022-05-05T19:58:47.858386Z ERROR panic: postgres stash connection error: db error: FATAL: terminating connection due to administrator command
thread: tokio:work-0
location: /home/pstoev/mjibson-coord-stash2/src/materialized/src/lib.rs:206:21
version: 0.26.1-dev (0602e93cd4e036665840ed72e47e882d321763a2)
backtrace:
   0: materialized::handle_panic
   1: core::ops::function::Fn::call
   2: std::panicking::rust_panic_with_hook
             at /rustc/7737e0b5c4103216d6fd8cf941b7ab9bdbaace7c/library/std/src/panicking.rs:702:17
   3: std::panicking::begin_panic_handler::{{closure}}
             at /rustc/7737e0b5c4103216d6fd8cf941b7ab9bdbaace7c/library/std/src/panicking.rs:588:13
   4: std::sys_common::backtrace::__rust_end_short_backtrace
             at /rustc/7737e0b5c4103216d6fd8cf941b7ab9bdbaace7c/library/std/src/sys_common/backtrace.rs:138:18
   5: rust_begin_unwind
             at /rustc/7737e0b5c4103216d6fd8cf941b7ab9bdbaace7c/library/std/src/panicking.rs:584:5
   6: core::panicking::panic_fmt
             at /rustc/7737e0b5c4103216d6fd8cf941b7ab9bdbaace7c/library/core/src/panicking.rs:143:14
   7: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
   8: tokio::runtime::task::harness::poll_future
   9: tokio::runtime::task::harness::Harness<T,S>::poll
  10: std::thread::local::LocalKey<T>::with
  11: tokio::runtime::thread_pool::worker::Context::run_task
  12: tokio::runtime::thread_pool::worker::Context::run
  13: tokio::macros::scoped_tls::ScopedKey<T>::set
  14: tokio::runtime::thread_pool::worker::run
  15: <tokio::runtime::blocking::task::BlockingTask<T> as core::future::future::Future>::poll
  16: tokio::runtime::task::harness::Harness<T,S>::poll
  17: tokio::runtime::blocking::pool::Inner::run
  18: std::sys_common::backtrace::__rust_begin_short_backtrace
  19: core::ops::function::FnOnce::call_once{{vtable.shim}}
  20: <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once
             at /rustc/7737e0b5c4103216d6fd8cf941b7ab9bdbaace7c/library/alloc/src/boxed.rs:1853:9
      <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once
             at /rustc/7737e0b5c4103216d6fd8cf941b7ab9bdbaace7c/library/alloc/src/boxed.rs:1853:9
      std::sys::unix::thread::Thread::new::thread_start
             at /rustc/7737e0b5c4103216d6fd8cf941b7ab9bdbaace7c/library/std/src/sys/unix/thread.rs:108:17
  21: start_thread
             at /build/glibc-eX1tMB/glibc-2.31/nptl/pthread_create.c:477:8
  22: clone
```

**Review by philip-stoev — APPROVED** [review]
I exercised this manually as much as I could and also temporarily converted the limits, the feature-benchmark and the upgrade test (this_version to this_version) to use postgres stash.

I will open a follow-up ticket for myself to push the updated tests as well as push a test that checks that pg restarts are gracefully handled, once that is fixed.

**Comment by mjibson** [comment]
:+1: I'm going to fix the issues listed here, then merge the last commit here. Probably some days.

**Comment by mjibson** [comment]
Researched item 5, the perf problem. It exists today on main, and doesn't seem to be a result of any stash stuff, so is unrelated to this PR. It occurs because the inner loop of rewrite_items (which is being run after the catalog is loaded) is fairly slow per item. So with hundreds or thousands of items, startup takes a while (it was 24 seconds with 400 tables, so easily many minutes with thousands).

**Comment by philip-stoev** [comment]
I am opening a ticket for item No 5.

**Comment by benesch** [comment]
Item #7 is #5797.

**Comment by mjibson** [comment]
I've read over item 8 a few times but don't understand what it's asking. Can you restate or show some sqlite database rows that illustrate your point in a different way?

**Comment by philip-stoev** [comment]
> I've read over item 8 a few times but don't understand what it's asking. Can you restate or show some sqlite database rows that illustrate your point in a different way?

Apologies. I meant that apparently you can not sort the entire `data` table by the `time` column and get a history of all the events that have been inserted. New stuff may be inserted that has a `time` value that is lower than the maximum `time` value in the table.

Here is an example.

1. Start a frezh mz instance
2. Naively attempt to dump the "most recent" event in the stash from sqlite:
```
sqlite> select time, length(key), length(value), collection_id from data desc order by time  limit 1;
-9223372036854775807|6|2|6
```

3. Create a database object of some sort in Mz:
```
materialize=> create table t1 (f1 integer);
CREATE TABLE
```

4. Go back to sqlite and attempt to observe the "most recent" event in the stash again:

```
sqlite> select time, length(key), length(value), collection_id from data desc order by time  limit 1;
-9223372036854775807|6|2|6
```

It is the same as before, as the event for `t1` was recorded under a time that is lower than `-9223372036854775807`.

So I was wondering, would that be a problem for any background jobs that attempt to compact the stash? I see that all keys defined have `collection_id` as a prefix, so I guess all operations will happen independently per `collection_id`.

**Comment by mjibson** [comment]
Ah, yes, that's by design. Each collection worries about its own timestamps, they are not related, even when committed as an append batch.

**Comment by mjibson** [comment]
I think all issues are fixed or pending, so I'm moving ahead with auto merging this.

**Comment by mjibson** [comment]
This was more complicated than I thought. Added a commit that moved a bunch of stuff to CatalogState, then a wrapper method that did the catalog copy and drop stuff for postgres or sqlite in a common way. DIsabling auto merge if anyone wants to look at those.

**Review by benesch — APPROVED** [review]
Didn't look _too_ closely, but seems good!

**Pull request merged**